### ISSUE TITLE
pkg_resources imported too early

### ISSUE THREAD

**Issue by jhamet93** [pantsbuild/pex#749]
I ran across two cases recently which resulted in pkg_resources being imported too early (resulting in the global working set being set before sys.path is done being finalized). This issue is also relevant to #713, #730, #732 .

1.  In setuptools, If a package sets `namespace_packages` to be an empty list in the setup.py, `has_metadata` will return True in the executed code below in environment.py.
```
def _declare_namespace_packages(resolved_dists):
    namespace_package_dists = [dist for dist in resolved_dists if dist.has_metadata('namespace_packages.txt')]
    if not namespace_package_dists:
      return  # Nothing to do here.
    pkg_resources, vendored = _import_pkg_resources()
    ...
```
However, since the list is empty, it is not actually a namespace package. This results in pkg_resources to be imported too early. I discovered this issue when using the package, `autopep8` in which the dependency, `pycodestyle`, was causing this.(https://github.com/PyCQA/pycodestyle/blob/2.5.0/setup.py). Therefore, to be defensive, should the code also ensure there's a value within the metadata ? 

2.  It seems the current logic is only cognizant of making sure not to import pkg_resources too early according to the executed pex.  Therefore, when setting PEX_PATH to merge in other pexes into the current environment, pkg_resources can be imported too early if there is a namespace package before the last merged pex. Should the code be modified to declare all the namespace packages only after all the pexes have been merged ? 

Both issues can be reproduced identically.
```
pex setuptools autopep8 -o environment.pex
pex six -o missing_dependency.pex
PEX_PATH=$(pwd)/missing_dependency.pex: ./environment.pex
```
```
>>> import pkg_resources
>>> pkg_resources.get_distribution('six')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/Users/jhamet/.pex/install/setuptools-41.0.1-py2.py3-none-any.whl.ffa66c57fd90eee9c21efeb1d3d918a7b960fa4f/setuptools-41.0.1-py2.py3-none-any.whl/pkg_resources/__init__.py", line 481, in get_distribution
    dist = get_provider(dist)
  File "/Users/jhamet/.pex/install/setuptools-41.0.1-py2.py3-none-any.whl.ffa66c57fd90eee9c21efeb1d3d918a7b960fa4f/setuptools-41.0.1-py2.py3-none-any.whl/pkg_resources/__init__.py", line 357, in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
  File "/Users/jhamet/.pex/install/setuptools-41.0.1-py2.py3-none-any.whl.ffa66c57fd90eee9c21efeb1d3d918a7b960fa4f/setuptools-41.0.1-py2.py3-none-any.whl/pkg_resources/__init__.py", line 900, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/Users/jhamet/.pex/install/setuptools-41.0.1-py2.py3-none-any.whl.ffa66c57fd90eee9c21efeb1d3d918a7b960fa4f/setuptools-41.0.1-py2.py3-none-any.whl/pkg_resources/__init__.py", line 786, in resolve
    raise DistributionNotFound(req, requirers)
DistributionNotFound: The 'six' distribution was not found and is required by the application
>>> import importlib
>>> importlib.reload(pkg_resources)
<module 'pkg_resources' from '/Users/jhamet/.pex/install/setuptools-41.0.1-py2.py3-none-any.whl.ffa66c57fd90eee9c21efeb1d3d918a7b960fa4f/setuptools-41.0.1-py2.py3-none-any.whl/pkg_resources/__init__.pyc'>
>>> pkg_resources.get_distribution('six')
six 1.12.0 (/Users/jhamet/.pex/install/six-1.12.0-py2.py3-none-any.whl.56417fecea1347c6c047607ab9996cc97aec99de/six-1.12.0-py2.py3-none-any.whl)
```

I can work on a pull request, just want to affirm these are issues.  Solving the second case should handle any issue brought forth from the first case.

**Comment by kwlzn** [comment]
confirming active repro on my machine:

```
[omerta xxx]$ pex --version
pex 1.6.8
[omerta xxx]$ pex setuptools autopep8 -o environment.pex
[omerta xxx]$ pex six -o missing_dependency.pex
[omerta xxx]$ PEX_PATH=$(pwd)/missing_dependency.pex ./environment.pex
Python 2.7.15 (default, Jun 20 2018, 17:40:09) 
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import pkg_resources; pkg_resources.get_distribution('six')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/Users/kwilson/.pex/install/setuptools-41.0.1-py2.py3-none-any.whl.ffa66c57fd90eee9c21efeb1d3d918a7b960fa4f/setuptools-41.0.1-py2.py3-none-any.whl/pkg_resources/__init__.py", line 481, in get_distribution
    dist = get_provider(dist)
  File "/Users/kwilson/.pex/install/setuptools-41.0.1-py2.py3-none-any.whl.ffa66c57fd90eee9c21efeb1d3d918a7b960fa4f/setuptools-41.0.1-py2.py3-none-any.whl/pkg_resources/__init__.py", line 357, in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
  File "/Users/kwilson/.pex/install/setuptools-41.0.1-py2.py3-none-any.whl.ffa66c57fd90eee9c21efeb1d3d918a7b960fa4f/setuptools-41.0.1-py2.py3-none-any.whl/pkg_resources/__init__.py", line 900, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/Users/kwilson/.pex/install/setuptools-41.0.1-py2.py3-none-any.whl.ffa66c57fd90eee9c21efeb1d3d918a7b960fa4f/setuptools-41.0.1-py2.py3-none-any.whl/pkg_resources/__init__.py", line 786, in resolve
    raise DistributionNotFound(req, requirers)
DistributionNotFound: The 'six' distribution was not found and is required by the application
>>> reload(pkg_resources)
<module 'pkg_resources' from '/Users/kwilson/.pex/install/setuptools-41.0.1-py2.py3-none-any.whl.ffa66c57fd90eee9c21efeb1d3d918a7b960fa4f/setuptools-41.0.1-py2.py3-none-any.whl/pkg_resources/__init__.pyc'>
>>> import pkg_resources; pkg_resources.get_distribution('six')
six 1.12.0 (/Users/kwilson/.pex/install/six-1.12.0-py2.py3-none-any.whl.56417fecea1347c6c047607ab9996cc97aec99de/six-1.12.0-py2.py3-none-any.whl)
>>> 
```

this is blocking our usage of various Google Cloud & Tensorflow related deps (in a `PEX_PATH` chain) without a forced `pkg_resources` reload.

**Comment by jhamet93** [comment]
Created a pull request with an integration test to address the issue: #750

**Comment by jsirois** [comment]
Fixed in #750
